hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/PrngSystemContract.java

```diff
@@ -0,0 +1,145 @@
+import static java.util.Objects.requireNonNull;
+import static org.hyperledger.besu.evm.frame.ExceptionalHaltReason.INVALID_OPERATION;
+
+import com.hedera.hapi.node.base.ContractID;
+import com.hedera.node.app.service.contract.impl.state.ProxyWorldUpdater;
+import com.hedera.node.app.service.contract.impl.utils.SystemContractUtils.ResultStatus;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.precompile.AbstractPrecompiledContract;
+import org.hyperledger.besu.evm.precompile.PrecompiledContract;
+
+/**
+ * System contract to generate random numbers. This will generate 256-bit pseudorandom number when
+ * no range is provided using n-3 record's running hash. If 32-bit integer "range" and 256-bit
+ * "seed" is provided returns a pseudorandom 32-bit integer X belonging to [0, range).
+ */
+@Singleton
+public class PrngSystemContract extends AbstractPrecompiledContract {
+    private static final Logger log = LogManager.getLogger(PrngSystemContract.class);
+    private static final String PRECOMPILE_NAME = "PRNG";
+    // random256BitGenerator(uint256)
+    static final int PSEUDORANDOM_SEED_GENERATOR_SELECTOR = 0xd83bf9a1;
+    public static final String PRNG_PRECOMPILE_ADDRESS = "0x169";
+    private long gasRequirement;
+
+    @Inject
+    public PrngSystemContract(@NonNull final GasCalculator gasCalculator) {
+        super(PRECOMPILE_NAME, gasCalculator);
+    }
+
+    @Override
+    public long gasRequirement(final Bytes bytes) {
+        return gasRequirement;
+    }
+
+    @Override
+    public @NonNull PrecompileContractResult computePrecompile(final Bytes input, @NonNull final MessageFrame frame) {
+        requireNonNull(input);
+        requireNonNull(frame);
+
+        // compute the gas requirement
+        gasRequirement =
+                calculateGas(Instant.ofEpochSecond(frame.getBlockValues().getTimestamp()));
+
+        // get the contract ID
+        final ContractID contractID = asEvmContractId(Address.fromHexString(PRNG_PRECOMPILE_ADDRESS));
+
+        try {
+            // compute the pseudorandom number
+            final var randomNum = generatePseudoRandomData(input, frame);
+            requireNonNull(randomNum);
+            final var result = PrecompiledContract.PrecompileContractResult.success(randomNum);
+
+            // create a child record
+            createSuccessfulRecord(frame, randomNum, contractID);
+
+            return result;
+        } catch (Exception e) {
+            log.warn("Internal precompile failure", e);
```

Comment:
This isn't exactly right (though it may be what mono service did!).  Because if the user called the contract at this address with a bad selector that's _on him_, right?  It's not an internal failure.  The internal failure is insufficient entropy (because you called this thing in the first three blocks).

Reply:
Your right.  Will rework.